### PR TITLE
2.1.0

### PR DESCRIPTION
**Pull request by iwatakeshi**
This update introduces themes into Doxx, and also the notifier. In addition, the default template has been separated into its own [repo](https://iwatakeshi/doxx-theme-default). This allows us to focus on designing the theme and not worry about the core.

I'll leave this pull request open before merging so I can get some feedback. Happy coding!

### CODE REVIEW

**Comment by iwatakeshi** [comment]
@FGRibreau This update should be safe to merge now, but if I can get some feedback it would be great.

Here's the wrap up of what happened:

* Installing themes is now possible using bower
* Forget about creating bower files or config files
  * You can easily install a theme by using `--theme <theme>` where `<theme>` is the name of the theme
    * The name of the theme can start with `doxx-theme-` but it doesn't have to
* It should be backward compatible with templates but I would advise users in future updates that templates will be deprecated...I'll leave that up to you.
* I added `update-notifier`for #78 but I haven't tested it. It should work because it was just a matter of copy and paste from the docs :stuck_out_tongue_closed_eyes:

File: README.md [hunk 1]
```diff
@@ -68,11 +68,11 @@ $ doxx --help
     -d, --debug                                Outputs parsed comments for debugging.
     -t, --title <string>                       The title for the page produced.
     -s, --source <source>                      The folder which should get parsed.
-    -i, --ignore <directories>                 Comma seperated list of directories to ignore. Default: test,public,static,views,templates
+    -i, --ignore <directories>                 Comma seperated list of directories to ignore. Default: test,public,static, view, view, views, templates
```
Comment (FGRibreau) [inline README.md]: one less `view`?

File: bin/doxx [hunk 2]
```diff
@@ -1,11 +1,22 @@
...
+var program         = require('commander'),
+    version         = require('../package').version,
+    doxx            = require('../src/doxx'),
+    option          = require('../src/option'),
+    _               = require('lodash'),
+    updateNotifier  = require('update-notifier'),
+    pkg             = require('../package.json');
+
+// Checks for available update and returns an instance
+var notifier = updateNotifier({pkg: pkg});
```
Comment (FGRibreau) [inline bin/doxx]: :+1:

**Comment by FGRibreau** [comment]
Hello @iwatakeshi ,

I will only have one feedback:

```shell
➜  doxx git:(iwatakeshi/master) ./bin/doxx --source lib --theme doxx-theme-cayman --target docs
doxx: Successfully installed theme: doxx-theme-cayman!
```

I think it would be better to [display a spinner](https://www.npmjs.com/package/elegant-spinner) when the theme is installing and to display a success message when the docs is written successfully.

PS: the cayman theme is really awesome!!

**Comment by cliftonc** [comment]
Afternoon @iwatakeshi - I tried to use this locally against my project, but ran into an issue where it refused to output anything.  

All folders created fine, but no actual HTML docs.

Steps:

```
git clone git@github.com:cliftonc/seguir.git
git clone git@github.com:iwatakeshi/doxx.git
cd doxx
npm install
gulp build
npm link
cd ../seguir
doxx --source api --target docs --theme doxx-theme-cayman
```

Docs folder is created, with the folders, but no html files?

Agree re. the spinner when installing a theme, but really looking forward to using this with the new cayman theme, it looks great.

**Comment by cliftonc** [comment]
Specifically, it is the following promise that isn't resolving:

https://github.com/iwatakeshi/doxx/blob/master/lib/theme.js#L305

File: README.md (hunk 1 above)
Comment (iwatakeshi) [inline README.md]: Fixed!

**Comment by iwatakeshi** [comment]
@FGRibreau @cliftonc :+1:  Thanks guys for the feedback. I'm checking out right now :) As far as the spinner is concerned, I'll see what library may be out there.

**Comment by iwatakeshi** [comment]
@cliftonc What the problem could be is that because doxx is stored globally, deleting a file isn't possible because of permissions? Does it give you an error output?

**Comment by cliftonc** [comment]
No error, it just doesn't resolve.  It is likely due to this:

https://github.com/sindresorhus/del/blob/master/index.js#L18

But the promise is swallowing the error perhaps?

**Comment by cliftonc** [comment]
With the above changes it now works for me, so many thanks.

Can I ask a question? I'm not sure why the theme management piece is downloading and then deleting the theme each time?

My very strong preference would be to store it somewhere - e.g. in ~/.doxx/themes - permanently, with an extra command to clean it / force re-download.

**Comment by iwatakeshi** [comment]
@cliftonc Very good question and it's good that you mentioned that because I didn't think of that haha! Things like this happen when one is a lone wolf but I'll see what I can do. Also a fix is coming up in cayman where some files can't find the assets.

**Comment by cliftonc** [comment]
Great - I'd be happy to submit it as a PR if you prefer once you have all these fixes in place with it working as is.

**Comment by iwatakeshi** [comment]
@cliftonc @FGRibreau  I fixed up the cache but just the command part in ./bin/doxx needs a fix. @cliftonc If you would like to fix that part, you're more than welcome to do so.

**Comment by iwatakeshi** [comment]
@cliftonc I forgot to mention that I basically wanted to add a command that:
* cleans the cache dir
* remove the cache dir
And that's about it :)

**Comment by iwatakeshi** [comment]
@FGRibreau @cliftonc OK, I think it's fixed up now. I'll see if I can fix up the theme and hopefully we can release this **awesome** docs generator. To be honest, even I am surprised how far I took it. :stuck_out_tongue_closed_eyes:
As usual, let me know if anything else is missing.

**Comment by cliftonc** [comment]
Looks good - but the CLI output is still saying that it is clearing the temp folder and downloading the theme each time?

**Comment by cliftonc** [comment]
One final issue - all of the file paths are still a little off.

For a file in a sub-folder from the ```--target```, it gets put into the HTML as 

```<li><a href="../auth/index.js.html">auth/index.js</a></li>```

It needs to be:

```<li><a href="auth/index.js.html">auth/index.js</a></li>```

And then if I go into that file ```auth/index.js```, the path for the static assets are now incorrect, they are:

```<link rel="stylesheet" href="css/bootstrap.css">```

They should be:

```<link rel="stylesheet" href="../css/bootstrap.css">```

You basically need to always re-state the path to the file, relative to the base (you can do this with Path.relative: https://nodejs.org/api/path.html#path_path_relative_from_to

**Comment by iwatakeshi** [comment]
@cliftonc Yup, I noticed that after testing it. I'm thinking it has to do with https://github.com/FGRibreau/doxx/blob/master/lib/doxx.js#L68-L98. I'm trying to figure out the code there because it works but at the same time it doesn't.

**Comment by iwatakeshi** [comment]
@cliftonc Could you check if it's working for you. On my end, the nested directories should be able to get assets now. Also, I update the log statements and should be more clear that bower will not download if a cached version is available. :)

**Comment by cliftonc** [comment]
@iwatakeshi that now works great, but you left a console.log in there:  https://github.com/iwatakeshi/doxx-theme-cayman/blob/0d1a4d01801bfc96c54cf53fc01d0c1f1dba8c1d/template/index.jade#L32

From my perspective this works, and is definitely a huge step forward for the project as a whole - once 2.1.0 lands I'll then be able to submit suggested improvements via PR vs just commenting on all the good work you are doing :)

**Comment by iwatakeshi** [comment]
@FGRibreau I'll change it up a bit :). @cliftonc I'm flattered, thanks :). I'm in the middle of doing additional minor changes to the API where it would allow developers who create awesome plugins (e.g. grunt, gulp) to use static methods instead of being restricted to Doxx specific stuff. Overall, It's still working good and the next update is coming soon.

**Comment by iwatakeshi** [comment]
@FGRibreau @cliftonc Further testing is needed but so far it looks good on my end.
@FGRibreau I wasn't sure if you wanted the private exactly like [this](https://github.com/iwatakeshi/doxx/blob/master/lib/doxx.js#L157-L193) but it does look cleaner to me.

**Comment by FGRibreau** [comment]
@iwatakeshi instead of a `var` + `_.forEach`, you can directly use a `return _.reduce` :)

**Comment by FGRibreau** [comment]
@iwatakeshi merge when you are ready and I'll make the release :)

**Comment by iwatakeshi** [comment]
@FGRibreau Thanks, I do try to think outside the box but I guess I don't use it too often haha. By the way, what line would that be? :)

**Comment by FGRibreau** [comment]
@iwatakeshi [there](https://github.com/iwatakeshi/doxx/blob/master/lib/doxx.js#L160-L162) :)

**Comment by iwatakeshi** [comment]
@FGRibreau Would you mind showing me how it works haha I'm not seeing how it's used yet. :) Checking out [this site](https://www.airpair.com/javascript/javascript-array-reduce) tells me that reduce will return a total and a variable which I'm going to assume that it will return `files:[]`is each `file:{}`.

**Comment by cliftonc** [comment]
May have meant a map?  You could replace the outer ``var targets = [] ... return targets;``` with:

```
    return _.map(this.files, (f) => {
        // code to figure out the path stuff
        return {
            name: f.name
        } // + other properties
    });
```

**Comment by iwatakeshi** [comment]
@FGRibreau @cliftonc I think it's ready but I'll commit it as a test update so we can test it and then from your feedbacks I'll merge it if it's good to go into the wild. :)

**Comment by FGRibreau** [comment]
@iwatakeshi my bad, indeed, `map` is enough here.

**Comment by iwatakeshi** [comment]
@FGRibreau @cliftonc I'm thinking it's all good now. Let me know if you encounter any issues before I merge this. :)

**Comment by cliftonc** [comment]
Looks good to me - can't wait for it to land so we can use it :)

**Comment by iwatakeshi** [comment]
@cliftonc @FGRibreau Alrighty, merging starts now! :)

**Comment by iwatakeshi** [comment]
:+1: :smile:  @cliftonc @FGRibreau Done! :) I'll leave the final touches to @FGRibreau .

**Comment by FGRibreau** [comment]
@iwatakeshi @cliftonc  https://github.com/FGRibreau/doxx/blob/master/CHANGELOG.md#v210-2015-09-14 :+1: !!

**Comment by iwatakeshi** [comment]
@FGRibreau My pleasure! I think it's a great project and wished it was more active :).

**Pull request merged**